---
title: "ptb_gwas"
author: "sq-96"
date: "2021-07-07"
output: workflowr::wflow_html
editor_options:
  chunk_output_type: console
---

```{r,echo=FALSE}
enrichment_test <- function(diff_pvalue_file, pip_cutoff, pvalue_cutoff){
  gene_pip <- read.table('./data/large_pipsumgenes.tsv',sep='\t',header=TRUE)
  gene_pip <- gene_pip[c('Symbol','pip_tot')]
  colnames(gene_pip) = c('gene','pip')
  
  diff_p <- read.table(paste('./data/differential_expression/',diff_pvalue_file,'.txt',sep = ''),sep='\t',header=TRUE)
  diff_p <- diff_p[c('gene','padj')]
  
  pip_p <- merge(gene_pip,diff_p,by='gene')
  pip_p[is.na(pip_p)] = 1
  pip_p <- pip_p[order(pip_p$pip,decreasing = TRUE),]
  
  high_sig <- sum((pip_p$pip>pip_cutoff) & (pip_p$padj<pvalue_cutoff))
  high_nosig <- sum((pip_p$pip>pip_cutoff) & (pip_p$padj>=pvalue_cutoff))
  nohigh_sig <- sum((pip_p$pip<=pip_cutoff) & (pip_p$padj<pvalue_cutoff))
  nohigh_nosig <- sum((pip_p$pip<=pip_cutoff) & (pip_p$padj>=pvalue_cutoff))
  
  sig_high_frac <- high_sig/(high_sig+high_nosig)
  sig_nohigh_frac <- nohigh_sig/(nohigh_sig+nohigh_nosig)
  
  ct <- matrix(c(high_sig,nohigh_sig,high_nosig,nohigh_nosig), ncol=2)
  result<-fisher.test(ct)
  return(c(diff_pvalue_file,round(result$estimate,3),result$p.value,round(result$conf.int[1],3),round(result$conf.int[2],3),sig_high_frac,sig_nohigh_frac))
}
```

## Enrichment test with pip cutoff = 0.3 and p-value cutoff = 0.1
```{r,echo=FALSE}
output_table <- data.frame()

pip_cutoff <- 0.3
pvalue_cutoff <- 0.1

output_table[1,1:5] <- enrichment_test('dec.PLXtcm.PL',pip_cutoff,pvalue_cutoff)[1:5]
output_table[2,1:5] <- enrichment_test('unt.PLXdec.PL',pip_cutoff,pvalue_cutoff)[1:5]
output_table[3,1:5] <- enrichment_test('unt.TLXdec.TL',pip_cutoff,pvalue_cutoff)[1:5]
output_table[4,1:5] <- enrichment_test('dec.TLXtcm.TL',pip_cutoff,pvalue_cutoff)[1:5]
output_table[5,1:5] <- enrichment_test('dec.PLXdec.TL',pip_cutoff,pvalue_cutoff)[1:5]
output_table[6,1:5] <- enrichment_test('unt.PLXctr.TL',pip_cutoff,pvalue_cutoff)[1:5]
# output_table[7,1:5] <- enrichment_test('tcm.PLXtcm.TL',pip_cutoff,pvalue_cutoff)[1:

colnames(output_table) = c('comparison','estimate','p.value','conf.low','conf.high')
output_table[order(output_table$estimate),]
```

```{r,message=FALSE,echo=FALSE,message=FALSE, warning=FALSE}
library(metaseqR)
enrichment_test_combined <- function(diff_pvalue_file1, diff_pvalue_file2, pip_cutoff, pvalue_cutoff){
  gene_pip <- read.table('./data/large_pipsumgenes.tsv',sep='\t',header=TRUE)
  gene_pip <- gene_pip[c('Symbol','pip_tot')]
  colnames(gene_pip) = c('gene','pip')
  
  diff_p_1 <- read.table(paste('./data/differential_expression/',diff_pvalue_file1,'.txt',sep = ''),sep='\t',header=TRUE)
  diff_p_1 <- diff_p_1[c('gene','padj')]
  
  diff_p_2 <- read.table(paste('./data/differential_expression/',diff_pvalue_file2,'.txt',sep = ''),sep='\t',header=TRUE)
  diff_p_2 <- diff_p_2[c('gene','padj')]
  
  diff_p <- cbind(diff_p_1$padj,diff_p_2$padj)
  diff_p <- fisher.method(diff_p,p.corr = "none")$p.adj
  diff_p <- data.frame(gene=as.character(diff_p_1$gene),padj=diff_p)
  
  pip_p <- merge(gene_pip,diff_p,by='gene')
  pip_p[is.na(pip_p)] = 1
  pip_p <- pip_p[order(pip_p$pip,decreasing = ),]
  
  high_sig <- sum((pip_p$pip>pip_cutoff) & (pip_p$padj<pvalue_cutoff))
  high_nosig <- sum((pip_p$pip>pip_cutoff) & (pip_p$padj>=pvalue_cutoff))
  nohigh_sig <- sum((pip_p$pip<=pip_cutoff) & (pip_p$padj<pvalue_cutoff))
  nohigh_nosig <- sum((pip_p$pip<=pip_cutoff) & (pip_p$padj>=pvalue_cutoff))
  
  sig_high_frac <- high_sig/(high_sig+high_nosig)
  sig_nohigh_frac <- nohigh_sig/(nohigh_sig+nohigh_nosig)
  
  ct <- matrix(c(high_sig,nohigh_sig,high_nosig,nohigh_nosig), ncol=2)
  result<-fisher.test(ct)
  return(c(paste(diff_pvalue_file1,diff_pvalue_file2),round(result$estimate,3),result$p.value,round(result$conf.int[1],3),round(result$conf.int[2],3),sig_high_frac,sig_nohigh_frac))
}
  
```

```{r,echo=FALSE}
dec.PL_unt.PL <- as.numeric(enrichment_test('unt.PLXdec.PL',pip_cutoff,pvalue_cutoff)[6:7])
dec.TL_unt.TL <- as.numeric(enrichment_test('unt.TLXdec.TL',pip_cutoff,pvalue_cutoff)[6:7])
dec.PL_tcm.PL <- as.numeric(enrichment_test('dec.PLXtcm.PL',pip_cutoff,pvalue_cutoff)[6:7])
dec.TL_tcm.TL <- as.numeric(enrichment_test('dec.TLXtcm.TL',pip_cutoff,pvalue_cutoff)[6:7])
#dec_tcm <- as.numeric(enrichment_test_combined('dec.PLXtcm.PL','dec.TLXtcm.TL',pip_cutoff,pvalue_cutoff)[6:7])
#ctr_dec <- as.numeric(enrichment_test_combined('unt.PLXdec.PL','unt.TLXdec.TL',pip_cutoff,pvalue_cutoff)[6:7])

df <- data.frame(group = rep(c("PIP > 0.3", "PIP < 0.3"),each=4),
                 comparison = rep(c("Dec. vs. Ctrl (PL)","Dec. vs. Ctrl (TL)","TCM vs. Dec. (PL)","TCM vs. Dec. (TL)"),2),
                 fraction = c(dec.PL_unt.PL[1],dec.TL_unt.TL[1],dec.PL_tcm.PL[1],dec.TL_tcm.TL[1],
                              dec.PL_unt.PL[2],dec.TL_unt.TL[2],dec.PL_tcm.PL[2],dec.TL_tcm.TL[2]))

df$comparison <- factor(df$comparison, levels = c("Dec. vs. Ctrl (PL)","Dec. vs. Ctrl (TL)","TCM vs. Dec. (PL)","TCM vs. Dec. (TL)"))
df$group <- factor(df$group, levels = c("PIP > 0.3", "PIP < 0.3"))
```

## Percent of DEGs in high PIP vs. low PIP genes
```{r,echo=FALSE,message=FALSE}
library(ggplot2)
p <- ggplot(data=df, aes(x=comparison, y=fraction, fill=group)) +
geom_bar(stat="identity", color="black", position=position_dodge())+theme_minimal()+ylab("Percent DEG")+xlab("Comparison")+guides(fill=guide_legend(title="Group"))
p + scale_fill_brewer(palette="Blues") + 
  annotate("rect", xmin = 0.75, xmax = 1.25, ymin = 0.7, ymax = 0.7, alpha=1,colour = "black")+
  annotate("rect", xmin = 1.75, xmax = 2.25, ymin = 0.58, ymax = 0.58, alpha=1,colour = "black")+
  annotate("rect", xmin = 2.75, xmax = 3.25, ymin = 0.45, ymax = 0.45, alpha=1,colour = "black")+
  annotate("rect", xmin = 3.75, xmax = 4.25, ymin = 0.4, ymax = 0.4, alpha=1,colour = "black")+
  annotate("text", x = 1, y = 0.72, label = "p = 0.022", size = 3.5)+
  annotate("text", x = 2, y = 0.60, label = "p = 0.112", size = 3.5)+
  annotate("text", x = 3, y = 0.47, label = "p = 0.001", size = 3.5)+
  annotate("text", x = 4, y = 0.42, label = "p = 0.014", size = 3.5)
```


```{r,echo=FALSE}
make_bubble_plot <- function(diff_pvalue_file, pip_cutoff, pvalue_cutoff){
  gene_pip <- read.table('./data/large_pipsumgenes.tsv',sep='\t',header=TRUE)
  gene_pip <- gene_pip[c('Symbol','pip_tot')]
  colnames(gene_pip) = c('gene','pip')
  
  diff_p <- read.table(paste('./data/differential_expression/',diff_pvalue_file,'.txt',sep = ''),sep='\t',header=TRUE)
  diff_p <- diff_p[c('gene','padj')]
  
  pip_p <- merge(gene_pip,diff_p,by='gene')
  pip_p[is.na(pip_p)] = 1
  pip_p <- pip_p[order(pip_p$pip,decreasing = TRUE),]
  pip_p$gene <- as.character(pip_p$gene)
  
  pip_p <- pip_p[pip_p$pip>0.3,]
  pip_p$padj <- -log10(pip_p$padj)
  #ggplot(pip_p, aes(x=pip,y=gene,size=padj))+geom_point(alpha=0.7)+labs(title = diff_pvalue_file)+theme(axis.text=element_text(size=16),axis.title=element_text(size=20),plot.title=element_text(size=20,hjust = 0.5))
  return(pip_p)
}
```

## DEG and PIP statistics of candidate genes (PIP > 0.3)
```{r,echo=FALSE,message=FALSE}
library(plot.matrix)
library(RColorBrewer)
df1 <- make_bubble_plot('dec.PLXdec.TL',pip_cutoff,pvalue_cutoff)
df2 <- make_bubble_plot('dec.PLXtcm.PL',pip_cutoff,pvalue_cutoff)
df3 <- make_bubble_plot('dec.TLXtcm.TL',pip_cutoff,pvalue_cutoff)
df4 <- make_bubble_plot('tcm.PLXtcm.TL',pip_cutoff,pvalue_cutoff)
df5 <- make_bubble_plot('unt.PLXctr.TL',pip_cutoff,pvalue_cutoff)
df6 <- make_bubble_plot('unt.PLXdec.PL',pip_cutoff,pvalue_cutoff)
df7 <- make_bubble_plot('unt.TLXdec.TL',pip_cutoff,pvalue_cutoff)
#df_heatmap <- cbind(df1$padj,df2$padj,df3$padj,df4$padj,df5$padj,df6$padj,df7$padj)
#rownames(df_heatmap) <- df1$gene
#colnames(df_heatmap) <- c("dec.PL_dec.TL","dec.PL_tcm.PL","dec.TL_tcm.TL","tcm.PL_tcm.TL","unt.PL_ctr.TL","unt.PL_dec.PL","unt.TL_dec.TL")
```

```{r,fig.height=10, fig.width=15,echo=FALSE}
#par(mar=c(5.1, 4.1, 4.1, 4.1))
#layout(mat = matrix(c(1, 2), nrow = 1),widths = c(1, 0.4),heights = c(1,0.5))
#p1 <- plot(as.matrix(df_heatmap),axis.col=list(las=1, cex.axis=1), axis.row=list(las=2, cex.axis=1),xlab="", ylab="", main="-log10(p-value)",col=brewer.pal(8,name = "Reds"),spacing.key=c(1,0.5,0))
#p2 <- plot(as.matrix(df1$pip),digits=3,key=NULL,col='white',axis.col=NULL, axis.row=NULL,xlab="", ylab="",main="PIP")
#p2 <- barplot(df1$pip[order(df1$pip,decreasing = TRUE)],horiz=TRUE)
```

```{r,echo=FALSE}
#df1["group"] <- "dec.PL_dec.TL"
#df2["group"] <- "dec.PL_tcm.PL"
#df3["group"] <- "dec.TL_tcm.TL"
#df4["group"] <- "tcm.PL_tcm.TL"
#df5["group"] <- "unt.PL_ctr.TL"
#df6["group"] <- "unt.PL_dec.PL"
#df7["group"] <- "unt.TL_dec.TL"


df2["group"] <- "TCM vs. Dec. (PL)"
df3["group"] <- "TCM vs. Dec. (TL)"
df6["group"] <- "Dec. vs. Ctrl (PL)"
df7["group"] <- "Dec. vs. Ctrl (TL)"
df_heatmap <- rbind(df2,df3,df6,df7)
df_heatmap$gene <- factor(df_heatmap$gene, levels = df1$gene)
df1$gene <- factor(df1$gene, levels = df1$gene)
df_heatmap$padj[which(df_heatmap$padj>6)]=6
```

```{r,fig.height=10, fig.width=15,echo=FALSE,message=FALSE,warning=FALSE,echo=FALSE}
library(gridExtra)
library(grid)
library(lattice)
p1 <- ggplot(data=df_heatmap,aes(x=group,y=gene,fill=padj))+geom_tile()+theme(panel.background = element_blank(),axis.ticks.y = element_blank(),axis.ticks.x = element_blank(),legend.position="left")+scale_fill_distiller(palette = "Blues",direction = 2,name='-log10-q-value')+labs(x ="", y = "")+guides(fill = guide_legend(label.position = "right", label.hjust = 1))+theme(plot.margin=margin(12,12,12,12),text = element_text(size=20))

p2 <- ggplot(data=df1, aes(x=gene, y=pip)) +
geom_bar(stat="identity", color="#fb9a99", position=position_dodge(),fill='#fb9a99')+coord_flip()+theme(panel.background = element_blank(),axis.text.y = element_blank(),axis.ticks.y = element_blank(), axis.line.x = element_line(),axis.title.y = element_blank())+theme(plot.margin=margin(12,12,12,12),text = element_text(size=20))

grid.arrange(p1,p2,ncol=2,nrow=1,widths=c(0.8, 0.2))
```

## Track Plot FOXL2
```{r,echo=FALSE,message=FALSE,warning=FALSE}
### Track plotting 
library(GenomicRanges) # required for all Gviz input
library(tidyverse) # manipulating tibbles
library(rtracklayer) # loading bigwigs/bed files
library(bigsnpr) # loading genotype data from 1000Genomes for LD calculation
library(Gviz) # main plotting
library(GenomicInteractions) # hic plots

setwd("/project2/xinhe/shengqian/ptb_gwas/")
source("./code/analysis_utils.R")
```

```{r,echo=FALSE}
merge.bigsnp.gwas <- function(gwas, bigSNP){
  
  map <- bigSNP$map
  snp_info <- map[,c('chromosome','physical.pos','allele1','allele2')]
  colnames(snp_info) <- c('chr','pos','a0','a1')
  
  matched.gwas <- as_tibble(bigsnpr::snp_match(gwas, 
                                               snp_info, 
                                               strand_flip = T, 
                                               match.min.prop = 0.1)) %>% dplyr::rename(og_index = `_NUM_ID_.ss`) %>% dplyr::rename(bigSNP_index = `_NUM_ID_`)
  
  return(matched.gwas)
}
```

```{r,echo=FALSE}
setwd('/project2/xinhe/shared_data/')
big.snp <- bigsnpr::snp_attach('/project2/xinhe/1kg/bigsnpr/EUR_variable_1kg.rds') # LD information
setwd('/project2/xinhe/shengqian/ptb_gwas/')
finemap_results_df <- readRDS('./data/snp_pip_gwas.RDS')
colnames(finemap_results_df) <- c('rsid','chr','pos','a0','a1','pval','pip','locus')
finemap_results_df$beta <- 0
finemap_results_df$pval <- -log10(finemap_results_df$pval)
finemap_results_df <- merge.bigsnp.gwas(finemap_results_df,big.snp)
```

```{r,echo=FALSE}
#rs17315501
#rs147843771
#rs62270785
LOCUS_TO_PLOT = 362
pip.df <- finemap_results_df[finemap_results_df$locus == LOCUS_TO_PLOT,] 
locus.chrom <- paste0("chr",pip.df$chr[1])
start.pos <- min(pip.df$pos)
end.pos <- max(pip.df$pos)

top.snp <- pip.df$bigSNP_index[which.max(pip.df$pval)]
top.snp.G <- big.snp$genotypes[,top.snp]
G.mat <- big.snp$genotypes[,pip.df$bigSNP_index]
r2.vals <- as.vector(cor(top.snp.G, G.mat))^2
r2.brackets <- cut(r2.vals, breaks = c(-1, 0.1, 0.25, 0.75, 0.9, 1), labels = c("0-0.1","0.1-0.25","0.25-0.75","0.75-0.9","0.9-1"))
pip.df$r2 <- r2.brackets
```

```{r,echo=FALSE}
pval.df <- pip.df[,c("chr","pos", "pval","r2")] %>% mutate(start = pos, end = pos) %>% dplyr::select(-pos) %>% pivot_wider(names_from = r2, values_from = "pval") 
pval.df.gr <- makeGRangesFromDataFrame(pval.df, keep.extra.columns = T)
seqlevelsStyle(pval.df.gr) <- "UCSC"

pval.track <- DataTrack(range = pval.df.gr,  genome = "hg19", groups = names(mcols(pval.df.gr)), col = c("black","blue","green","red","orange"), name = "-log10 pvalue")

## Finemap PIP track
pip.track <- DataTrack(data = pip.df$pip, chromosome = pip.df$chr, start = pip.df$pos, end = pip.df$pos, genome = "hg19", name = "PIP")

# axis track - so we know where we are in the genome 
axisTrack <- GenomeAxisTrack()
```

```{r,echo=FALSE}
atac.PT.TCM <- rtracklayer::import('./data/PL-TCM.bw')
atac.PT.TCM.track <- DataTrack(range = atac.PT.TCM, type = "h", genome = "hg19", name = "ATAC.PT.TCM", col = "navy", showAxis=FALSE, ylim = c(0, 2.21))

#atac.PT.dec <- rtracklayer::import('./data/PL-dec.bw')
#atac.PT.dec.track <- DataTrack(range = atac.PT.dec, type = "h", genome = "hg19", name = "ATAC.PT.dec", col = "navy", showAxis=FALSE, #ylim = c(0, 2.21))

#atac.TL.TCM <- rtracklayer::import('./data/TL-TCM.bw')
#atac.TL.TCM.track <- DataTrack(range = atac.TL.TCM, type = "h", genome = "hg19", name = "ATAC.TL.TCM", col = "navy", showAxis=FALSE, ylim = c(0, 1))

#atac.TL.dec <- rtracklayer::import('./data/TL-dec.bw')
#atac.TL.dec.track <- DataTrack(range = atac.TL.dec, type = "h", genome = "hg19", name = "ATAC.TL.dec", col = "navy", showAxis=FALSE, ylim = c(0, 1))

```

```{r,echo=FALSE}
curr.locus.gr <- GRanges(seqnames = paste0("chr",pip.df$chr[1]), IRanges(start =  start.pos, 
                                                                             end = end.pos ))
gene.track <- knownGeneObject(curr.locus.gr = curr.locus.gr, genome = "hg19")
```

```{r,echo=FALSE}
get_gene_name <- function(raw_name){
  index <- which(strsplit(raw_name,"")[[1]] == "*")[1]
  gene_name <- substr(raw_name, start=1, stop = index-1)
  return(gene_name)
}
```

```{r,echo=FALSE}
pcHic <- read.table("./data/chicago-dPL5-TCM.ibed",header = TRUE)
pcHic$gene_name <- sapply(pcHic$bait_name,FUN = get_gene_name)
enhancer.pcHiC.gr <- GRanges(seqnames = pcHic$bait_chr, ranges = IRanges(start = pcHic$bait_start, end = pcHic$bait_end), score = pcHic$score, gene = pcHic$gene_name)
promoter.pcHiC.gr <- GRanges(seqnames = pcHic$otherEnd_chr, ranges = IRanges(start = pcHic$otherEnd_start, end =pcHic$otherEnd_end), score = pcHic$score)
pchic.obj <- GenomicInteractions::GenomicInteractions(anchor1 = enhancer.pcHiC.gr, anchor2 = promoter.pcHiC.gr)
pchic.obj$counts <- round(pchic.obj$anchor1.score)
```

```{r,echo=FALSE}
pchic.obj.filt <- pchic.obj[which(pchic.obj$anchor1.gene == "FOXL2"),]
hic.track <- InteractionTrack(pchic.obj.filt, name = "pcHiC.PL.TCM")

dpars <- list(col.interactions="red", 
              col.anchors.fill ="black", 
              col.anchors.line = "black", 
              interaction.dimension = "height",  
              interaction.measure = "counts", 
              plot.trans = FALSE, 
              plot.outside = FALSE,  
              col.outside="lightblue",  
              anchor.height = 0.1,
              interaction.dimension.transform = "log10")
displayPars(hic.track) <- dpars
```

```{r,echo=FALSE}
snp.p1 = 139029676
snp.p2 = 138843356
list.of.tracks <- c()
ht1 <- HighlightTrack(c(pval.track, pip.track, atac.PT.TCM.track, gene.track, hic.track, axisTrack), 
                          start = c(snp.p1,snp.p2)-500, width = 1000, 
                          chromosome = locus.chrom, col = 'pink')
```

```{r,echo=FALSE}
# actually plot
plotTracks(ht1, # tracks in a list
           chromosome = locus.chrom,            # the chromosome we want
           from = start.pos + 1200000,         # the start of the region we want
           to = end.pos - 700000,             # the end of the region
           sizes = c(1, 1, 0.3, 0.5, 1, 1), # relative size of each track
           panel.only = F,          # do you want labels or just the pure panel? change to True to see what i mean
           collapseTranscripts= 'longest',
           transcriptAnnotation = "symbol",
           col.title = "black",
           background.title="transparent", 
           col.axis="black", 
           col.title="black"
)
```

